### PR TITLE
Add messagepack encoding

### PR DESCRIPTION
**Pull request by PabloDelBarrioArnanz**
PR to add messagepack encoding for rabbit message, this functionality can be activated indicating that the contentType of the message is 'application/x-msgpack' any other type must work as usual.

Json2msgpack library is used for marshalling.

### CODE REVIEW

**Review by javaducky — CHANGES_REQUESTED** [review]
Sorry for the extreme delay...this was lost in my Inbox.

I'm not so sure we should accept this as the `github.com/izinin/json2msgpack` does not have any licensing information associated with it.

**Comment by pablochacin** [comment]
Hi @PabloDelBarrioArnanz I'm not sure that adding explicit support for content type encoding in a transport extension is the best approach. Imagine how [this code](https://github.com/grafana/xk6-amqp/pull/17/files#diff-fd8340611142499a9c6547a1fcabce57bf8dd435629a1ef89276bcfa896dc4c4R86-R90) would evolve if we keep adding support for more content types:
```
	if options.ContentType == messagepack {
		publishing.Body = json2msgpack.EncodeJSON([]byte(options.Body))
	} else {
		publishing.Body = []byte(options.Body)
	}
```

I'm therefore wondering why not make the encoding/decoding in JS either using an existing JS library (in a quick search I found [this](https://github.com/msgpack/msgpack-javascript)) or by implementing the msgpack encoding/decoding in its own extension (say xk6-msgpack)

**Comment by PabloDelBarrioArnanz** [comment]
> 



> Sorry for the extreme delay...this was lost in my Inbox.
> 
> I'm not so sure we should accept this as the `github.com/izinin/json2msgpack` does not have any licensing information associated with it.

Hello @javaducky!!
True, I did not notice that it does not have a defined license, maybe if I used this implementation, would you accept the feature? 😄
=> https://github.com/vmihailenco/msgpack

**Comment by PabloDelBarrioArnanz** [comment]
> Hi @PabloDelBarrioArnanz I'm not sure that adding explicit support for content type encoding in a transport extension is the best approach. Imagine how [this code](https://github.com/grafana/xk6-amqp/pull/17/files#diff-fd8340611142499a9c6547a1fcabce57bf8dd435629a1ef89276bcfa896dc4c4R86-R90) would evolve if we keep adding support for more content types:
> 
> ```
> 	if options.ContentType == messagepack {
> 		publishing.Body = json2msgpack.EncodeJSON([]byte(options.Body))
> 	} else {
> 		publishing.Body = []byte(options.Body)
> 	}
> ```
> 
> I'm therefore wondering why not make the encoding/decoding in JS either using an existing JS library (in a quick search I found [this](https://github.com/msgpack/msgpack-javascript)) or by implementing the msgpack encoding/decoding in its own extension (say xk6-msgpack)

Hello namesake 😄!!

About the implementation is true, understanding what you mean. A concatenation of if-else for each content-type would be horrible.

But as to whether it is the right place to use the ContentType variable, I think so. As far as I know text/plain, application/json, application/msgpack... are data exchange format.

Regarding whether to the encoding in the test itself with js, I think it would not be optimal since anyone who wants to do a test of this type would have to import the library and duplicate code between tests.
>

**Comment by PabloDelBarrioArnanz** [comment]
> > maybe if I used this implementation, would you accept the feature? 😄 => https://github.com/vmihailenco/msgpack
> 
> @PabloDelBarrioArnanz Yes, if you were to switch the backing implementation used for the compression, we could accept the PR.

Hi @javaducky, switched! 😊

File: amqp.go [hunk 1]
```diff
@@ -78,7 +83,22 @@ func (amqp *AMQP) Publish(options PublishOptions) error {
...
 	publishing := amqpDriver.Publishing{
 		ContentType: options.ContentType,
-		Body:        []byte(options.Body),
+	}
+
+	if options.ContentType == messagepack {
+		var jsonParsedBody interface{}
+		json.Unmarshal([]byte(options.Body), &jsonParsedBody)
+
+		if err := json.Unmarshal([]byte(options.Body), &jsonParsedBody); err != nil {
```
Comment (javaducky) [inline amqp.go:92]:
```suggestion
		if err = json.Unmarshal([]byte(options.Body), &jsonParsedBody); err != nil {
```

File: amqp.go [hunk 2]
```diff
@@ -78,7 +83,22 @@ func (amqp *AMQP) Publish(options PublishOptions) error {
 
 	publishing := amqpDriver.Publishing{
 		ContentType: options.ContentType,
-		Body:        []byte(options.Body),
+	}
+
+	if options.ContentType == messagepack {
+		var jsonParsedBody interface{}
+		json.Unmarshal([]byte(options.Body), &jsonParsedBody)
```
Comment (javaducky) [inline amqp.go:90]: Delete this...it appears unmarshaling is happening twice.

**Review by javaducky — CHANGES_REQUESTED** [review]

File: amqp.go (hunk 2 above)
Comment (PabloDelBarrioArnanz) [reply amqp.go]: Sorry, absolutely right, it seems that I'm blind... 😅

File: amqp.go (hunk 1 above)
Comment (PabloDelBarrioArnanz) [reply amqp.go]: Sure doesn't need to be declared again

**Review by javaducky — APPROVED** [review]
LGTM! Thank you for your contribution and attention to this.

**Pull request merged**